pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -199,11 +210,25 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(env *m
 		vHostPortMap[virtualHostWrapper.Port] = append(vHostPortMap[virtualHostWrapper.Port], virtualHosts...)
 	}
 
-	var virtualHosts []*route.VirtualHost
+	var tmpVirtualHosts []*route.VirtualHost
 	if listenerPort == 0 {
-		virtualHosts = mergeAllVirtualHosts(vHostPortMap)
+		tmpVirtualHosts = mergeAllVirtualHosts(vHostPortMap)
+	} else {
+		tmpVirtualHosts = vHostPortMap[listenerPort]
+	}
+
+	var virtualHosts []*route.VirtualHost
```

Comment:
@yxue / @howardjohn the pilot CPU increase comes from this function in general. In the situation with 300 listeners and 300 RDS routes, we are computing the SAME rds for all 300 and then pruning that RDS. Thats where the CPU bump is coming from, as we are computing this very large RDS over and over again. Unfortunately, this fix will be a bit more involved as the current RDS code assumes that we always do RDS for 0.0.0.0:PORT, instead of x.x.x.x:PORT - different semantics. So, one option is to build the RDS once, cache it somewhere (probably in the config gen or push context?). The only issue is that we are doing proxy version specific config generation.

Reply:
Ah I see that makes sense, I didn't get a chance to really look at the code yet. Ideally we actually just inline the routes in the new listeners instead of even using RDS (if it is feasible/more efficient)?

Reply:
We cant as we would end up draining the entire listener or filter chain as opposed to just updating RDS which is cheaper.